### ISSUE TITLE
uhm yes hello these modules are amazing, thank you 

### ISSUE THREAD

**Issue by gwatcha** [The-Modular-Mind/oscelot#5]

**Comment by The-Modular-Mind** [comment]
Hehe ty, feel free to suggest improvements/features. Also if you have OSC templates/presets for whatever software you use with VCV + OSC'elot and want to share

**Comment by gwatcha** [comment]
yeah! that would be fun. 

I've been using this OpenStageControl template for some time now on Rack 1 with trowasoft's osc modules. 

![image](https://user-images.githubusercontent.com/31526638/146632517-6a67476d-1cdd-4440-b406-1f6fd7b1c203.png)

here's a demo from the depths of the sea
https://www.twitch.tv/videos/1236570371

(I also have an expression pedal and two foot buttons to control the Circle module on the left. It's my own module I've been working on for quite some time now, it handles all the sound emergence.)

I'm still working on setting it up in Rack v2.
I just got the cat to receive messages from OpenStageControl. I still gotta port my module though. 
(For any OpenStageControl users reading - I set the ending of `address `with `/fader` and set the `preArgs `to the fader numbers ('1', '2'...)  (I noticed that it crashes otherwise - one suggestion is to possibly catch it and print out 'format error'))

**Comment by gwatcha** [comment]
Pairing OSC with VCV Rack has really uphauled my music workflow to be based around live creation (+ Circle). I was very relieved when I found your modules in the library when I sat down to get v2 working today. Then I started playing around with them and I was like, wow, there is a lot of care put into these. So thanks a lot again 🙏 .

**Comment by gwatcha** [comment]
I got it all working today and it's great. One thing I love is how it remembers the parameters when I boot up Rack.

It looks great
![image](https://user-images.githubusercontent.com/31526638/146660452-d3ca85f2-2a57-4a74-a868-b5654615e4e2.png)

here's my previous 'howls moving castle' patch
![image](https://user-images.githubusercontent.com/31526638/146660486-55b840e8-b224-4aba-8471-5afa3ad4655b.png)

😌

**Comment by The-Modular-Mind** [comment]
Hmm how they all have the same fader IDs, are you using one fader to control all of them, or is there a bug  there

**Comment by gwatcha** [comment]
There's a bug. It works somehow though. The messages I send are of this format:

e.g. 1
address: `/delay/mix/fader`
preArgs: 0

e.g. 2
address: `/delay/feedback/fader`
preArgs: 0

You see I set the preArgs to 0 for everyone, because I noticed it didn't make a difference, as they were differentiated anyways, and I was lazy.

Also, when I set the preArg to '1' (as well as '2' & higher), I got a long number like FDR-16497724..

Reading how OpenStageControl works, it says `preArgs` is - 'a value that will be prepended to the widget's value in the OSC messages it sends'. I wish I knew how to see the message I am sending so I can show you it for certain.

**Comment by The-Modular-Mind** [comment]
Ah I see, ok yes I did code it so that it'll work as long as the address/id combo is distinct.

All OSC messages consist of an address and a list of args.

OSC'elot expects 2 arguments, if you give more its fine, will be ignored as long as it has a usable ID and value.

In OpenStageControl pre-args are just 'hardcoded' args. The other args come from moving the slider like value. Your config is fine :)

**Comment by The-Modular-Mind** [comment]
The issue with the long fader ID seems like a bug, are you setting the type of ID to Int and value to Float. Should be type 'if' in OpenStageControl. You can also see a basic OpenStageControl preset in the  OSC'elot folder.

**Comment by gwatcha** [comment]
Ok great. Ah yes it seems like it is cast to float by default. Setting the 'if' there fixed it. I did not know of the preset's, that would have helped I'm sure (I'm not sure if you do mention it in the manual, if you do that's good).

**Comment by gwatcha** [comment]
What do you use?

**Comment by The-Modular-Mind** [comment]
Yeah I think I've mentioned it, if not I will add it when I update it with the new features, banks etc.

I use a hardware controller and OpenStage for testing:
https://community.vcvrack.com/t/do-we-still-need-the-real-thing/12018/123

**Comment by gwatcha** [comment]
Ah cool. It's nice you have something tactile.

**Comment by gwatcha** [comment]
You are on spaceship levels my good sir

**Comment by gwatcha** [comment]
Hey I noticed another issue that I have to sometimes open the patch twice in order for OSC'elot to recall ALL the mappings I made. Sometimes it only recalls the first page. It also has trouble turning on the receiver sometimes, and I have to restart for it to work.

I'm noticing I have to sometimes open a patch 3-4 times for both the receiver to work and OSC'elot to remember all the mappings. It is quite strange.

**Comment by The-Modular-Mind** [comment]
Hmm can you send me the patch? There might also be errors in the log when it doesn't connect.

**Comment by gwatcha** [comment]
https://drive.google.com/file/d/1E3dW6rSTQLxBBJoIGG8uGdtVOE8xtYAp/view?usp=sharing

Here you are.

Here is what I just did, maybe you can reproduce:

Open Rack, OSC'elot is missing many assignments, receiver starts green.
Re-open saved patch, OSC'elot finds missing assignments, receiver starts orange.
Receiver does not boot when button pressed.
Close Rack and re-run.
OSC'elot is missing assignments again, but receiver button turns on receiver.
Open saved patch, OSC'elot finds missing assignments, receiver does not boot when pressed however.
Close Rack and re-run.
OSC'elot is missing assignments again, receiver starts orange, do not try to boot.
Open saved patch. 
OSC'elot has assignments, and receiver boots when pressed.

Thank you for helping too :pray:

**Comment by gwatcha** [comment]
I just realized that if I press 'scan for next mapping' the mappings I thought were lost reappear. Though I would like them to all be mapped at the same time and it doesn't seem possible in this scan mode.

**Comment by The-Modular-Mind** [comment]
Sorry I was on a couple of weeks break from work/tech. 

- So the docs still need to be updated, but it is accurate for most things. Only the new features have to be added.
- The next/prev buttons are for switching between modules that you have saved mappings for. You can use the apply button to directly recall a mapping by clicking on a module. This is similar to Midi-Cat if you have used it. The docs have a some GIFs showing this. This only saves one module at a time.
- I've added a new save mode in V2, which saves all mappings in the current bank. To use this, you simply move the slider to the next bank and there are 128 banks. So you can use this to save multiple modules in one go.  So if you have some params mapped from multiple modules in bank 1, move the slider to 2, it'll save those in position 1 and give you a blank bank in Position 2 and so on.

Hope that clears up any problems/bugs, if you still see something lmk and I'll try to repro/fix

**Comment by gwatcha** [comment]
Thank you! I will.
